src/response_sender.cc

```diff
@@ -54,15 +72,32 @@ ResponseSender::~ResponseSender()
 }
 
 void
-ResponseSender::Send(
-    std::shared_ptr<InferResponse> infer_response, const uint32_t flags)
+ResponseSender::UpdateStateAndCounters(
+    const std::shared_ptr<InferResponse>& response, const uint32_t flags)
 {
-  // Release the GIL. This avoids a potential deadlock situation in the parent
-  // process, where every thread in the thread pool is indirectly waiting for a
-  // function in the stub process that acquires the GIL. Meanwhile, the current
-  // thread, which holds the GIL, is also waiting for the parent side to have
-  // the next available thread to pick up the job during resource contention.
-  py::gil_scoped_release release;
+  if (is_decoupled_ == nullptr) {
+    // TODO: Can a model access the response sender on a BLS infer request?
+    throw PythonBackendException(
+        "Unable to send response. Response sender has no reference to the "
+        "decoupled state of the model.");
+  }
+  bool is_decoupled = *is_decoupled_;
+
+  std::lock_guard<std::mutex> lk(mu_);
+
+  if (!is_decoupled) {
+    if (response != nullptr && number_of_response_sent_ > 0) {
+      throw PythonBackendException(
```

Comment:
We need to have a ticket to add some checks in the core as well.

Reply:
Sure, added the ticket.